### ISSUE TITLE
8-schools failing to compile (RSTAN: 2.19.2)

### ISSUE THREAD

**Issue by saptarshiguha** [stan-dev/rstan#700]
#### Summary:

When running the 8 schools example, it fails to compile

#### Description:
```
fit <- stan(file = '~/tmp/8schools.stan', data = schools_dat)
```
throws an error
```
Error in compileCode(f, code, language = language, verbose = verbose) :
  Compilation ERROR, function(s)/method(s) not created! In file included from /home/sguha/software/R_LIBS/BH/include/boost/random/detail/integer_log2.hpp:19:0,
                 from /home/sguha/software/R_LIBS/BH/include/boost/random/detail/int_float_pair.hpp:26,
                 from /home/sguha/software/R_LIBS/BH/include/boost/random/exponential_distribution.hpp:27,
                 from /home/sguha/software/R_LIBS/BH/include/boost/random/gamma_distribution.hpp:25,
                 from /home/sguha/software/R_LIBS/StanHeaders/include/stan/math/prim/mat/prob/dirichlet_rng.hpp:5,
                 from /home/sguha/software/R_LIBS/StanHeaders/include/stan/math/prim/mat.hpp:292,
                 from /home/sguha/software/R_LIBS/StanHeaders/include/stan/math/rev/mat.hpp:12,
                 from /home/sguha/software/R_LIBS/StanHeaders/include/stan/math.hpp:4,
                 from /home/sguha/software/R_LIBS/StanHeaders/include/src/stan/model/model_header.hpp:4,
                 from file291075013335.cpp:8:
/home/sguha/software/R_LIBS/BH/include/boost/pending/integer_log2.hpp:7:89: note: #pragma message: This header is deprecated. Use <boost/integer/integer_log2.hpp> instead.
 BOOST_HEADER_DEPRECATED("<boost/integer/integer_log2.hpp>");
                                                                                         ^
In file included from /home/sguha/software/R_LIBS/StanHeaders/include/stan/math/rev/mat/fun/ordered_constrain.hpp:6:0,
                 from /home/sguha/software/R_LIBS/StanHeaders/include/stan/math/rev/mat.hpp:41,
                 from /home/sguha/software/R_LIBS/StanHeaders/include/stan/math.hpp:4,
                 from /home/sguha/software/R_LIBS/StanHeaders/include/src/stan/model/model_header.hpp:4,
                 from file291075013335.cpp:8:
/home/sguha/software/R_LIBS/StanHeaders/include/stan/math/rev/mat/functor/adj_jac_apply.hpp:31:32: error: ‘std::index_sequence’ has not been declared
                           std::index_sequence<I...> i) {
                                ^
/home/sguha/software/R_LIBS/StanHeaders/include/stan/math/rev/mat/functor/adj_jac_apply.hpp:31:46: error: expected ‘,’ or ‘...’ before ‘<’ token
                           std::index_sequence<I...> i) {
                                              ^
/home/sguha/software/R_LIBS/StanHeaders/include/stan/math/rev/mat/functor/adj_jac_apply.hpp: In function ‘constexpr auto stan::math::internal::apply(const F&, const Tuple&)’:
/home/sguha/software/R_LIBS/StanHeaders/include/stan/math/rev/mat/functor/adj_jac_apply.hpp:49:27: error: ‘make_index_sequence’ is not a member of ‘std’
   return apply_impl(f, t, std::make_index_sequence<std::tuple_size<Tuple>{}>{});
/home/sguha/software/R_LIBS/StanHeaders/include/stan/math/rev/mat/functor/adj_jac_apply.hpp:49:27: note: suggested alternative:
In file included from /home/sguha/software/R_LIBS/BH/include/boost/fusion/container/vector/vector.hpp:28:0,
                 from /home/sguha/software/R_LIBS/BH/include/boost/fusion/container/vector/vector10.hpp:25,
                 from /home/sguha/software/R_LIBS/BH/include/boost/fusion/view/transform_view/transform_view.hpp:22,
                 from /home/sguha/software/R_LIBS/BH/include/boost/fusion/algorithm/transformation/transform.hpp:11,
                 from /home/sguha/software/R_LIBS/BH/include/boost/fusion/view/zip_view/detail/begin_impl.hpp:14,
                 from /home/sguha/software/R_LIBS/BH/include/boost/fusion/view/zip_view/zip_view.hpp:16,
                 from /home/sguha/software/R_LIBS/BH/include/boost/fusion/view/zip_view.hpp:12,
                 from /home/sguha/software/R_LIBS/BH/include/boost/fusion/include/zip_view.hpp:11,
                 from /home/sguha/software/R_LIBS/BH/include/boost/numeric/odeint/util/resize.hpp:26,
                 from /home/sguha/software/R_LIBS/BH/include/boost/numeric/odeint/util/state_wrapper.hpp:26,
                 from /home/sguha/software/R_LIBS/BH/include/boost/numeric/odeint/util/ublas_wrapper.hpp:33,
                 from /home/sguha/software/R_LIBS/BH/include/boost/numeric/odeint.hpp:25,
                 from /home/sguha/software/R_LIBS/StanHeaders/include/stan/math/prim/arr/functor/integrate_ode_rk45.hpp:17,
                 from /home/sguha/software/R_LIBS/StanHeaders/include/stan/math/prim/arr.hpp:46,
                 from /home/sguha/software/R_LIBS/StanHeaders/include/stan/math/prim/mat.hpp:344,
                 from /home/sguha/software/R_LIBS/StanHeaders/include/stan/math/rev/mat.hpp:12,
                 from /home/sguha/software/R_LIBS/StanHeaders/include/stan/math.hpp:4,
                 from /home/sguha/software/R_LIBS/StanHeaders/include/src/stan/model/model_header.hpp:4,
                 from file291075013335.cpp:8:
/home/sguha/software/R_LIBS/BH/include/boost/fusion/support/detail/index_sequence.hpp:59:12: note:   ‘boost::fusion::detail::make_index_sequence’
     struct make_index_sequence
            ^
In file included from /home/sguha/software/R_LIBS/StanHeaders/include/stan/math/rev/mat/fun/ordered_constrain.hpp:6:0,
                 from /home/sguha/software/R_LIBS/StanHeaders/include/stan/math/rev/mat.hpp:41,
                 from /home/sguha/software/R_LIBS/StanHeaders/include/stan/math.hpp:4,
                 from /home/sguha/software/R_LIBS/StanHeaders/include/src/stan/model/model_header.hpp:4,
                 from file291075013335.cpp:8:
/home/sguha/software/R_LIBS/StanHeaders/include/stan/math/rev/mat/functor/adj_jac_apply.hpp:49:77: error: expected primary-expression before ‘{’ token
   return apply_impl(f, t, std::make_index_sequence<std::tuple_size<Tuple>{}>{});
                                                                             ^
/home/sguha/software/R_LIBS/StanHeaders/include/stan/math/rev/mat/functor/adj_jac_apply.hpp:49:77: error: expected ‘)’ before ‘{’ token
/home/sguha/software/R_LIBS/StanHeaders/include/stan/math/rev/mat/functor/adj_jac_apply.hpp: At global scope:
/home/sguha/software/R_LIBS/StanHeaders/include/stan/math/rev/mat/functor/adj_jac_apply.hpp:151:9: error: expected type-specifier
       = std::result_of_t<F(decltype(is_var_), decltype(value_of(Targs()))...)>;
         ^
/home/sguha/software/R_LIBS/StanHeaders/include/stan/math/rev/mat/functor/adj_jac_apply.hpp:156:42: error: ‘FReturnType’ was not declared in this scope
   std::array<int, internal::compute_dims<FReturnType>::value> M_;
                                          ^
/home/sguha/software/R_LIBS/StanHeaders/include/stan/math/rev/mat/functor/adj_jac_apply.hpp:156:53: error: template argument 1 is invalid
   std::array<int, internal::compute_dims<FReturnType>::value> M_;
                                                     ^
/home/sguha/software/R_LIBS/StanHeaders/include/stan/math/rev/mat/functor/adj_jac_apply.hpp:156:61: error: template argument 2 is invalid
   std::array<int, internal::compute_dims<FReturnType>::value> M_;
                                                             ^
/home/sguha/software/R_LIBS/StanHeaders/include/stan/math/rev/mat/functor/adj_jac_apply.hpp: In member function ‘std::vector<stan::math::var> stan::math::adj_jac_vari<F, Targs>::build_return_varis_and_vars(co
nst std::vector<double>&)’:
/home/sguha/software/R_LIBS/StanHeaders/include/stan/math/rev/mat/functor/adj_jac_apply.hpp:349:9: error: invalid types ‘int[int]’ for array subscript
     M_[0] = val_y.size();
                                ^
/home/sguha/software/R_LIBS/StanHeaders/include/stan/math/rev/mat/functor/adj_jac_apply.hpp: In member function ‘Eigen::Matrix<stan::math::var, R, C> stan::math::adj_jac_vari<F, Targs>::build_return_varis_and_vars(const Eigen::Matrix<double, R, C>&)’:
/home/sguha/software/R_LIBS/S
```

Output of sessionInfo. Note, i uninstalled rstan, StanHeaders and reinstalled them.

```
> sessionInfo()
R version 3.5.0 (2018-04-23)
Platform: x86_64-redhat-linux-gnu (64-bit)
Running under: CentOS Linux 7 (Core)

Matrix products: default
BLAS/LAPACK: /usr/lib64/R/lib/libRblas.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C
 [9] LC_ADDRESS=C               LC_TELEPHONE=C
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base

other attached packages:
[1] brms_2.10.0        Rcpp_1.0.2         rstan_2.19.2       ggplot2_3.2.1
[5] StanHeaders_2.19.0
```

**Comment by saptarshiguha** [comment]
I should point out i have version  BH_1.69.0-1 of the BH library.

**Comment by reikoch** [comment]
Exactly the same problem here with rstan 2.19.2 on RHEL 7.6:
```
Error in compileCode(f, code, language = language, verbose = verbose) :
  Compilation ERROR, function(s)/method(s) not created! In file included from /opt/bee_tools/R/3.6.1/lib64/R/library_sec/BH/include/boost/random/detail/integer_log2.hpp:19:0,
                 from /opt/bee_tools/R/3.6.1/lib64/R/library_sec/BH/include/boost/random/detail/int_float_pair.hpp:26,
                 from /opt/bee_tools/R/3.6.1/lib64/R/library_sec/BH/include/boost/random/exponential_distribution.hpp:27,
                 from /opt/bee_tools/R/3.6.1/lib64/R/library_sec/BH/include/boost/random/gamma_distribution.hpp:25,
                 from /opt/bee_tools/R/3.6.1/lib64/R/library_sec/StanHeaders/include/stan/math/prim/mat/prob/dirichlet_rng.hpp:5,
                 from /opt/bee_tools/R/3.6.1/lib64/R/library_sec/StanHeaders/include/stan/math/prim/mat.hpp:292,
                 from /opt/bee_tools/R/3.6.1/lib64/R/library_sec/StanHeaders/include/stan/math/rev/mat.hpp:12,
                 from /opt/bee_tools/R/3.6.1/lib64/R/library_sec/StanHeaders/include/stan/math.hpp:4,

In addition: Warning message:
In system(cmd, intern = !verbose) :
  running command '/opt/bee_tools/R/3.6.1/lib64/R/bin/R CMD SHLIB filecf82c3f54f6.cpp 2> filecf82c3f54f6.cpp.err.txt' had status 1
Error in sink(type = "output") : invalid connection
```
here system info:
```
> sessionInfo()
R version 3.6.1 (2019-07-05)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Red Hat Enterprise Linux

Matrix products: default
BLAS/LAPACK: /usr/lib64/libopenblas-r0.3.3.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C
 [9] LC_ADDRESS=C               LC_TELEPHONE=C
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base

other attached packages:
[1] rstan_2.19.2       ggplot2_3.2.1      StanHeaders_2.19.0

loaded via a namespace (and not attached):
 [1] Rcpp_1.0.2         magrittr_1.5       tidyselect_0.2.5   munsell_0.5.0
 [5] colorspace_1.4-1   R6_2.4.0           rlang_0.4.0        dplyr_0.8.3
 [9] tools_3.6.1        parallel_3.6.1     pkgbuild_1.0.6     grid_3.6.1
[13] gtable_0.3.0       loo_2.1.0          cli_1.1.0          withr_2.1.2
[17] matrixStats_0.55.0 lazyeval_0.2.2     assertthat_0.2.1   tibble_2.1.3
[21] crayon_1.3.4       processx_3.4.1     gridExtra_2.3      purrr_0.3.2
[25] callr_3.3.2        ps_1.3.0           inline_0.3.15      glue_1.3.1
[29] compiler_3.6.1     pillar_1.4.2       prettyunits_1.0.2  scales_1.0.0
[33] stats4_3.6.1       pkgconfig_2.0.3
```
Is the compiler too old?
```
]$ gcc --version
gcc (GCC) 4.8.5 20150623 (Red Hat 4.8.5-39)
```

**Comment by bgoodri** [comment]
Yes, g++ 4.8.x is too old. Get
https://developers.redhat.com/products/developertoolset/overview
